{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Preliminary Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "This document includes some preliminary results based on the simulated data (from Joey):\n",
    "- Output of all files from _afflicted_test.zip_\n",
    "- Summary statistics: allele frequencies\n",
    "- Basic association analysis on the disease trait for all single SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "Simulation Data:\n",
    "- Obtained from https://www.synapse.org/#!Synapse:syn22314894\n",
    "- **Base population**: 4000 pedigrees, with 2-6 children generated by R package \"msprime\" (coalescent simulation method)\n",
    "    - 2-6 children reflects the actual distirbution of the number of children within a typical US family\n",
    "- **Whole genome**: 12500 variants, one arbitary chromosome\n",
    "- Offsprings generated from recombination via a segmented loop that randomly selects a parents haplotype via chunks\n",
    "    - The segments are 125 bp switches for the loop generating the offsprings\n",
    "    - Every 125 bp a parent's haplotype is randomly chosen and each child gets their segment, which allows for children not to be identical (GRM confirmed).\n",
    "- Each selected family must have >=2 individuals with the binary phenotype (V11)\n",
    "    - After selection (>=2 cases), there are **5680 individuals** and **710 families**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/duz/Desktop/Center_for_Statistical_Genetics/family-association_DZ/analysis\n"
     ]
    }
   ],
   "source": [
    "# Current working directory\n",
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [],
   "source": [
    "# Change working directory to simulation data folder\n",
    "cd ./afflicted_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs10085\n",
      "rs10119\n",
      "rs10120\n",
      "rs10207\n",
      "rs10233\n",
      "rs10235\n",
      "rs10269\n",
      "rs10366\n",
      "rs10368\n",
      "rs10605\n"
     ]
    }
   ],
   "source": [
    "# 1) my_snps.txt: Selected SNPs that were given effects (MAF around 0.1)\n",
    "head my_snps.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "kernel": "Bash",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs10085\t0.405465108108164\n",
      "rs10119\t0.405465108108164\n",
      "rs10120\t0.405465108108164\n",
      "rs10207\t0.405465108108164\n",
      "rs10233\t0.405465108108164\n",
      "rs10235\t0.405465108108164\n",
      "rs10269\t0.405465108108164\n",
      "rs10366\t0.405465108108164\n",
      "rs10368\t0.405465108108164\n",
      "rs10605\t0.405465108108164\n"
     ]
    }
   ],
   "source": [
    "# 2) snp_eff.txt: Each snp was pulled from the generated data with its effect size\n",
    "# Protective variant effect size = -0.693\n",
    "# Susceptible variant effect size = 0.405\n",
    "head snp_eff.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [],
   "source": [
    "# 3) GRM file: six_fam_eff_affected_grm.sXX.txt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "kernel": "Bash",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHR_A\tBP_A\tSNP_A\tCHR_B\tBP_B\tSNP_B\tR2\n",
      "1\t10085\trs10085\t1\t1\trs1\t3.47977e-05\n",
      "1\t10085\trs10085\t1\t2\trs2\t5.82632e-05\n",
      "1\t10085\trs10085\t1\t3\trs3\t5.49087e-05\n",
      "1\t10085\trs10085\t1\t4\trs4\t0.000318135\n",
      "1\t10085\trs10085\t1\t5\trs5\t0.000367395\n",
      "1\t10085\trs10085\t1\t6\trs6\t3.86758e-05\n",
      "1\t10085\trs10085\t1\t7\trs7\t8.34403e-05\n",
      "1\t10085\trs10085\t1\t8\trs8\t8.59001e-06\n",
      "1\t10085\trs10085\t1\t9\trs9\t8.59001e-06\n"
     ]
    }
   ],
   "source": [
    "# 4) SNPs not in LD with each selected SNPs\n",
    "# Quality control check\n",
    "head snps_not_in_ld_with_selected.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "kernel": "Bash",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\tid_3_1\n",
      "2\tid_4_1\n",
      "3\tid_5_1\n",
      "3\tid_6_1\n",
      "9\tid_17_1\n",
      "9\tid_18_1\n",
      "10\tid_19_1\n",
      "10\tid_20_1\n",
      "2\tid_22_1\n",
      "3\tid_23_1\n"
     ]
    }
   ],
   "source": [
    "# 5) six_fam_eff_sample_affected_fams.txt\n",
    "# After subsetting by afflicted and the appropriate distribution of families\n",
    "# FID and IID for >=2 cases and distributed families according to census info:\n",
    "head six_fam_eff_sample_affected_fams.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IID\tFID\tpop_id\tV1\tV11\n",
      "id_3_1\t2\t1\t3.52914085178346\t1\n",
      "id_4_1\t2\t1\t1.78691807704925\t1\n",
      "id_5_1\t3\t1\t-1.26760779020004\t1\n",
      "id_6_1\t3\t1\t2.27087740878414\t1\n",
      "id_17_1\t9\t1\t5.17571703602393\t0\n",
      "id_18_1\t9\t1\t-2.38575755858843\t0\n",
      "id_19_1\t10\t1\t2.43594972380419\t1\n",
      "id_20_1\t10\t1\t5.5551344722583\t1\n",
      "id_22_1\t2\t1\t5.15389011434599\t0\n"
     ]
    }
   ],
   "source": [
    "# 6) six_fam_eff_phenotypes_structured_selected_fams.txt\n",
    "# pop_id: population id\n",
    "# V1: Continuous trait\n",
    "# V11: Binary trait (0,1)\n",
    "head six_fam_eff_phenotypes_structured_selected_fams.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.52914085178346\t1\n",
      "1.78691807704925\t1\n",
      "-1.26760779020004\t1\n",
      "2.27087740878414\t1\n",
      "5.17571703602393\t0\n",
      "-2.38575755858843\t0\n",
      "2.43594972380419\t1\n",
      "5.5551344722583\t1\n",
      "5.15389011434599\t0\n",
      "0.515583680261195\t0\n"
     ]
    }
   ],
   "source": [
    "# 7) six_fam_eff_phenotypes_structured_one_selected_fams.txt\n",
    "# Same as 6), except only V1 and V11 results are included\n",
    "head six_fam_eff_phenotypes_structured_one_selected_fams.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr\trs\tps\tn_miss\tallele1\tallele0\taf\tbeta\tse\tlogl_H1\tl_remle\tl_mle\tp_wald\tp_lrt\tp_score\n",
      "1\trs2\t2\t0\tC\tA\t0.025\t-8.035558e-02\t1.913898e-01\t-1.109916e+04\t5.403260e+00\t5.403422e+00\t6.746088e-01\t6.745518e-01\t6.745773e-01\n",
      "1\trs4\t4\t0\tC\tA\t0.128\t6.971148e-02\t1.536459e-01\t-1.109915e+04\t5.404674e+00\t5.404446e+00\t6.500511e-01\t6.499817e-01\t6.500029e-01\n",
      "1\trs7\t7\t0\tC\tA\t0.265\t-2.162183e-02\t9.290714e-02\t-1.109922e+04\t5.403634e+00\t5.404270e+00\t8.159832e-01\t8.159403e-01\t8.159494e-01\n",
      "1\trs8\t8\t0\tC\tA\t0.145\t1.230853e-01\t1.271866e-01\t-1.109878e+04\t5.401501e+00\t5.401260e+00\t3.332088e-01\t3.331300e-01\t3.332241e-01\n",
      "1\trs9\t9\t0\tC\tA\t0.145\t1.230853e-01\t1.271866e-01\t-1.109878e+04\t5.401501e+00\t5.401260e+00\t3.332088e-01\t3.331300e-01\t3.332241e-01\n",
      "1\trs10\t10\t0\tC\tA\t0.226\t-9.550721e-03\t8.990818e-02\t-1.109924e+04\t5.403859e+00\t5.404329e+00\t9.154056e-01\t9.153798e-01\t9.153852e-01\n",
      "1\trs14\t14\t0\tC\tA\t0.128\t6.971148e-02\t1.536459e-01\t-1.109915e+04\t5.404674e+00\t5.404446e+00\t6.500511e-01\t6.499817e-01\t6.500029e-01\n",
      "1\trs15\t15\t0\tC\tA\t0.405\t4.720208e-02\t1.091661e-01\t-1.109916e+04\t5.405148e+00\t5.404858e+00\t6.654760e-01\t6.654180e-01\t6.654432e-01\n",
      "1\trs16\t16\t0\tC\tA\t0.065\t-7.040791e-02\t1.300597e-01\t-1.109910e+04\t5.402155e+00\t5.402156e+00\t5.882872e-01\t5.882442e-01\t5.883091e-01\n"
     ]
    }
   ],
   "source": [
    "# 8) six_fam_eff.assoc.txt\n",
    "# Basic association test results with Gemma and GRM\n",
    "head six_fam_eff.assoc.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [],
   "source": [
    "# 9) Plink file: six_fam_eff_affected.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\trs1\t0\t1\tC\tA\n",
      "1\trs2\t0\t2\tC\tA\n",
      "1\trs3\t0\t3\tC\tA\n",
      "1\trs4\t0\t4\tC\tA\n",
      "1\trs5\t0\t5\tC\tA\n",
      "1\trs6\t0\t6\tC\tA\n",
      "1\trs7\t0\t7\tC\tA\n",
      "1\trs8\t0\t8\tC\tA\n",
      "1\trs9\t0\t9\tC\tA\n",
      "1\trs10\t0\t10\tC\tA\n"
     ]
    }
   ],
   "source": [
    "# 10) Plink file: six_fam_eff_affected.bim\n",
    "head six_fam_eff_affected.bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 id_3_1 0 0 1 1\n",
      "2 id_4_1 0 0 1 1\n",
      "3 id_5_1 0 0 1 1\n",
      "3 id_6_1 0 0 1 1\n",
      "9 id_17_1 0 0 1 1\n",
      "9 id_18_1 0 0 1 1\n",
      "10 id_19_1 0 0 1 1\n",
      "10 id_20_1 0 0 1 1\n",
      "2 id_22_1 0 0 1 1\n",
      "3 id_23_1 0 0 1 1\n"
     ]
    }
   ],
   "source": [
    "# 11) Plink file: six_fam_eff_affected.fam\n",
    "# Columns: family individual_ID father mother sex affected_status(1=control, 2=case)\n",
    "# only FID and IID are useful variables; all the rest are dummy variables for plink import format\n",
    "head six_fam_eff_affected.fam "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "### Summary statistics: allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to freq_stat.log.\n",
      "Options in effect:\n",
      "  --1\n",
      "  --bfile six_fam_eff_affected\n",
      "  --freq\n",
      "  --out freq_stat\n",
      "  --pheno six_fam_eff_phenotypes_structured_selected_fams.txt\n",
      "\n",
      "8192 MB RAM detected; reserving 4096 MB for main workspace.\n",
      "12500 variants loaded from .bim file.\n",
      "5680 people (5680 males, 0 females) loaded from .fam.\n",
      "0 phenotype values present after --pheno.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5680 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "--freq: Allele frequencies (founders only) written to freq_stat.frq .\n"
     ]
    }
   ],
   "source": [
    "plink --bfile six_fam_eff_affected --pheno six_fam_eff_phenotypes_structured_selected_fams.txt --1 --freq --out freq_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR     SNP   A1   A2          MAF  NCHROBS\n",
      "   1     rs1    C    A     0.002113    11360\n",
      "   1     rs2    C    A      0.02535    11360\n",
      "   1     rs3    C    A            0    11360\n",
      "   1     rs4    C    A       0.1282    11360\n",
      "   1     rs5    C    A    0.0007042    11360\n",
      "   1     rs6    C    A     0.006338    11360\n",
      "   1     rs7    C    A       0.2651    11360\n",
      "   1     rs8    C    A       0.1447    11360\n",
      "   1     rs9    C    A       0.1447    11360\n"
     ]
    }
   ],
   "source": [
    "head freq_stat.frq "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "### Basic association analysis on the disease trait for all single SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "kernel": "Bash",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to as1.log.\n",
      "Options in effect:\n",
      "  --assoc\n",
      "  --bfile six_fam_eff_affected\n",
      "  --out as1\n",
      "\n",
      "8192 MB RAM detected; reserving 4096 MB for main workspace.\n",
      "12500 variants loaded from .bim file.\n",
      "5680 people (5680 males, 0 females) loaded from .fam.\n",
      "5680 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5680 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "12500 variants and 5680 people pass filters and QC.\n",
      "Among remaining phenotypes, 0 are cases and 5680 are controls.\n",
      "Writing C/C --assoc report to as1.assoc ... 1624324049576573819098done.\n"
     ]
    }
   ],
   "source": [
    "plink --bfile six_fam_eff_affected --assoc --out as1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "### Adjusted for multiple testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "kernel": "Bash",
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to as2.log.\n",
      "Options in effect:\n",
      "  --adjust\n",
      "  --assoc\n",
      "  --bfile six_fam_eff_affected\n",
      "  --out as2\n",
      "\n",
      "8192 MB RAM detected; reserving 4096 MB for main workspace.\n",
      "12500 variants loaded from .bim file.\n",
      "5680 people (5680 males, 0 females) loaded from .fam.\n",
      "5680 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 5680 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "12500 variants and 5680 people pass filters and QC.\n",
      "Among remaining phenotypes, 0 are cases and 5680 are controls.\n",
      "Writing C/C --assoc report to as2.assoc ... 1624324049576573819098done.\n",
      "--adjust: Genomic inflation est. lambda (based on median chisq) = 1.\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899--adjust values (11073 variants) written to as2.assoc.adjusted .\n"
     ]
    }
   ],
   "source": [
    "plink --bfile six_fam_eff_affected --assoc --adjust --out as2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "kernel": "Bash",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR     SNP         BP   A1      F_A      F_U   A2        CHISQ            P           OR \n",
      "   1     rs1          1    C       NA 0.002113    A            0            1           NA \n",
      "   1     rs2          2    C       NA  0.02535    A            0            1           NA \n",
      "   1     rs3          3    C       NA        0    A           NA           NA           NA \n",
      "   1     rs4          4    C       NA   0.1282    A            0            1           NA \n",
      "   1     rs5          5    C       NA 0.0007042    A            0            1           NA \n",
      "   1     rs6          6    C       NA 0.006338    A            0            1           NA \n",
      "   1     rs7          7    C       NA   0.2651    A            0            1           NA \n",
      "   1     rs8          8    C       NA   0.1447    A            0            1           NA \n",
      "   1     rs9          9    C       NA   0.1447    A            0            1           NA \n"
     ]
    }
   ],
   "source": [
    "head as2.assoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR     SNP      UNADJ         GC       BONF       HOLM   SIDAK_SS   SIDAK_SD     FDR_BH     FDR_BY\n",
      "   1 rs12500          1          1          1          1          1          1          1          1 \n",
      "   1     rs2          1          1          1          1          1          1          1          1 \n",
      "   1     rs4          1          1          1          1          1          1          1          1 \n",
      "   1     rs5          1          1          1          1          1          1          1          1 \n",
      "   1     rs6          1          1          1          1          1          1          1          1 \n",
      "   1     rs7          1          1          1          1          1          1          1          1 \n",
      "   1     rs8          1          1          1          1          1          1          1          1 \n",
      "   1     rs9          1          1          1          1          1          1          1          1 \n",
      "   1    rs10          1          1          1          1          1          1          1          1 \n"
     ]
    }
   ],
   "source": [
    "head as2.assoc.adjusted"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     ""
    ]
   ],
   "panel": {
    "displayed": true,
    "height": 0
   },
   "version": "0.21.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
